src/main/java/core/basesyntax/MyHashMap.java

```diff
@@ -1,24 +1,145 @@
+                    lastBucketNode.value = value;
+                    return;
+                }
+                lastBucketNode = lastBucketNode.nextNode;
+            }
+            lastBucketNode.nextNode = new Node<>(key.hashCode(), key, value, null);
+        }
+        size++;
+    }
+
+    private void putForNullKey(V value) {
+        if (bucketStock[0] == null) {
+            bucketStock[0] = new Node<>(0, null, value, null);
+        } else {
+            Node<K, V> lastBucketNode = bucketStock[0];
+            if (lastBucketNode.key == null) {
+                lastBucketNode.value = value;
+                return;
+            }
+            while (lastBucketNode.nextNode != null) {
+                if (lastBucketNode.key == null) {
+                    lastBucketNode.value = value;
+                    return;
+                }
+                lastBucketNode = lastBucketNode.nextNode;
+            }
+            lastBucketNode.nextNode = new Node<K, V>(0, null, value, null);
+        }
+        size++;
     }
 
     @Override
     public V getValue(K key) {
+        if (key == null) {
+            return getForNullKey();
+        }
+
+        int index = getIndexFromHash(getHash(key.hashCode()));
+        if (index >= bucketStock.length || bucketStock[index] == null) {
+            return null;
+        }
+        Node<K, V> returnNode = bucketStock[index];
+        while (returnNode != null) {
+            if (returnNode.key.equals(key)) {
+                return returnNode.value;
+            }
+            returnNode = returnNode.nextNode;
+        }
         return null;
     }
 
+    private V getForNullKey() {
+        Node<K, V> returnNode = bucketStock[0];
+        while (returnNode != null) {
+            if (returnNode.key == null) {
+                return returnNode.value;
+            }
+            returnNode = returnNode.nextNode;
+        }
+        return null;
+    }
+
+    private int getHash(int hashCode) {
+        hashCode ^= (hashCode >>> 20) ^ (hashCode >>> 12);
+        return hashCode ^ (hashCode >>> 7) ^ (hashCode >>> 4);
+
+    }
+
+    private int getIndexFromHash(int hash) {
+        return Math.abs(hash % bucketStock.length);
+    }
+
+    private void resize() {
+        if (size >= bucketStock.length * LOAD_FACTOR) {
+            Node<K, V>[] oldBucketStock = bucketStock;
+            bucketStock = new Node[bucketStock.length * 2];
+            //transfer
+            for (int i = 0; i < oldBucketStock.length; i++) {
+                if (bucketStock[i] == null) {
+                    continue;
+                }
+                while (bucketStock[i] != null) {
+                    put(bucketStock[i].key, bucketStock[i].value);
+                    bucketStock[i] = bucketStock[i].nextNode;
+                }
+            }
+        }
+    }
+
     @Override
     public int getSize() {
-        return 0;
+        return size;
+    }
+
+    public static class Node<K, V> {
+        private int hash;
```

Comment:
why do we save hash code here? where do we use it?

Reply:
true, nowhere, thank you)